File: Java/622. Design Circular Queue.java
class MyCircularQueue {

    private int capacity;
    private int[] queue;
    private int head;
    private int cur;
    private int count;
    
    public MyCircularQueue(int k) {
        capacity = k;
        queue = new int[k];
        head = 0;
        cur = -1;
        count = 0;
    }
    
    public boolean enQueue(int value) {
        // if queue is full
        if(count == capacity){
            return false;
        }
        
        // if the cur pointer reaches the end, we'll start over
        // else, we'll increment cur
        if(cur+1 == capacity){
            cur = 0;
        }else{
            cur++;
        }
        
        queue[cur] = value;
        count++;
        return true;
    }
    
    public boolean deQueue() {
        if(count == 0){
            return false;
        }
        
        // if the queue will be empty after this, refresh the vars
        if(count == 1){
            head = 0;
            cur = -1;
        }else{
            if(head == capacity-1){
                head = 0;
            }else{
                head++;
            }  
        }
        count--;
        return true;
    }
    
    public int Front() {
        return count == 0 ? -1 : queue[head];
    }
    
    public int Rear() {
        return count == 0 ? -1 : queue[cur];
    }
    
    public boolean isEmpty() {
        return count == 0;
    }
    
    public boolean isFull() {
        return count == capacity;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: Python3/622. Design Circular Queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.k = k
        self.q = [None]*self.k
        self.head = 0
        self.cur = -1
        self.count = 0
        #print("-------------------Start Over----------------")

    def enQueue(self, value: int) -> bool:
        
        # if queue is full, return False
        if self.count == self.k:
            return False
          
        # if the self.cur pointer reaches the end, we'll start over
        # else, we'll increment self.cur
        if self.cur+1 == self.k:
            self.cur = 0
        else:
            self.cur += 1
            
        self.q[self.cur] = value
        
        #print("enQueue: \t", self.head, self.cur, self.q)
        self.count += 1
        return True

    def deQueue(self) -> bool:
        # if the queue currently has no element
        if self.count == 0:
            return False
        
        # make the head pointer null
        self.q[self.head] = None
        
        # if the queue will be empty after this, refresh the vars
        if self.count == 1:
            self.head = 0
            self.cur = -1
            
        else:
            if self.head == self.k-1:
                self.head = 0
            else:
                self.head += 1
        self.count -= 1
           
        #print("deQueue: \t", self.head, self.cur, self.q)
        return True
        
    def Front(self) -> int:
        if self.count == 0:
            return -1
        #print("Front: \t", self.q[self.head])
        return self.q[self.head]

    def Rear(self) -> int:
        if self.count == 0:
            return -1
        #print("Rear: \t", self.q[self.cur])
        return self.q[self.cur]


    def isEmpty(self) -> bool:
        return self.count == 0

    def isFull(self) -> bool:
        return self.count == self.k


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()